### ISSUE TITLE
Cop requiring for space around interpolation

### ISSUE THREAD

**Issue by dziulius** [bbatsov/rubocop#1971]
There is a cop that forbids adding space around string interpolation #1836

It would be good if there was an opposite cop, which **requires** space around string interpolation:
```ruby
# Good
"some #{ interpolated } string"

# Bad
"some #{interpolated} string"
```

**Comment by jonas054** [comment]
Good addition of functionality IMO. But it should be implemented as a configuration option for `SpaceInsideStringInterpolation` rather than as a new cop:

```yaml
Style/SpaceInsideStringInterpolation:
  EnforcedStyle: no_space
  SupportedStyles:
    - space
    - no_space
```

**Comment by bbatsov** [comment]
Yep, you're totally right.

**Comment by sauliusgrigaitis** [comment]
I'm wondering what is best default for this cop. I personally prefer without spaces, because it eliminates any possible confusion and it's more closer to interpolated string comparing to version with spaces around variable. Any reasons why it would be better to use spaces around variable as default?

**Comment by bquorning** [comment]
I can't recall having seen spaces around interpolated strings in any Ruby book, blog post or source code. In my opinion, RuboCop should not need to support this particular style.

**Comment by sauliusgrigaitis** [comment]
@bquorning I think so too, but some of guys in my team prefers spaces around, so I'm trying to find constructive arguments why spaces around could be used. The only one argument I heard is "it looks more clean/nice with spaces around, similar to ruby hash'es or blocks".

**Comment by jonas054** [comment]
@sauliusgrigaitis That's a perfectly valid argument, but [the string interpolation section of The Ruby Style Guide](https://github.com/bbatsov/ruby-style-guide#string-interpolation) prescribes no spaces so that has to be the default (until someone gets a pull request that changes this recommendation merged into the style guide).

**Comment by philoserf** [comment]
The ruby community, based upon code publicly available, prefers no spaces. It isn't even close. We have looked at the question before. I will look for the threads.

**Comment by philoserf** [comment]
@sauliusgrigaitis. My position with team members advocating variant is that we default to the community standard unless and until we have substantial reason to diverge from the standard. It works for us. Part of my argument is that the community standard has the embedded wisdom of many. Having and using the standard makes interacting with the community much easier. I go on with more arguments than these as needed to ensure our variance is substantiated by our need not just desire.

**Comment by bbatsov** [comment]
Yeah, the default will remain `no_spaces`. There's little point in the discussing this right now, it's more important to implement the alternative config first.

**Comment by bquorning** [comment]
> it's more important to implement the alternative config first.

As I recall it, you have often commented that RuboCop shouldn’t (and cannot) support all different style preferences. As mentioned in my earlier comment above, I have never seen spaces around interpolated strings anywhere, and I don’t think there’s any need for the RuboCop codebase to support this style.

For rare styles like this, I’d rather advocate for writing a custom cop that you then load [as an extension](https://github.com/bbatsov/rubocop#extensions). I am doing that in at least one place where RuboCop and I disagree :-)

**Comment by dziulius** [comment]
Even rubocop sorce contains spaces around string interpolation https://github.com/bbatsov/rubocop/blob/master/spec/rubocop/cop/style/unneeded_capital_w_spec.rb#L14

**Comment by sauliusgrigaitis** [comment]
Rails also has a lot of places where interpolation is with spaces, like this https://github.com/rails/rails/blob/master/activesupport/test/core_ext/object/to_param_test.rb#L7 .

I personally prefer without spaces for variables interpolation, but what about one-liners that are interpolated, like this https://github.com/rails/rails/blob/master/railties/lib/rails/generators.rb#L181 ?

**Comment by bquorning** [comment]
> Even rubocop source contains spaces around string interpolation

Touché. Obviously, my argument doesn’t hold.

**Comment by bbatsov** [comment]
> As I recall it, you have often commented that RuboCop shouldn’t (and cannot) support all different style preferences. As mentioned in my earlier comment above, I have never seen spaces around interpolated strings anywhere, and I don’t think there’s any need for the RuboCop codebase to support this style.

This is something which I'm fine with supporting. It's not super uncommon.

**Comment by bbatsov** [comment]
> Even rubocop source contains spaces around string interpolation

Haha. Guess you found a bug in the cop. :-)

**Comment by sauliusgrigaitis** [comment]
That's not a bug. It's lack of dogfooding in rubocop project.

**Comment by bbatsov** [comment]
> That's not a bug. It's lack of dogfooding in rubocop project.

RuboCop is run on RuboCop's code for every PR; that's why I'm sure this is a bug.

**Comment by sauliusgrigaitis** [comment]
Hehe. That's good discussion. So if rubocop contributors actually doesn't follow that rule in style guide (obviously for some reason contributor used spaces) and no one noticed that for long time, then maybe it's something wrong with this rule.

**Comment by bbatsov** [comment]
Humans are unreliable reviewers. Everyone makes mistakes and misses things from time to time. This is why we have RuboCop. :-)

**Comment by sauliusgrigaitis** [comment]
Humans and RuboCop are both unreliable as its already proven above :) But that's not my main point.

I'm questioning this rule in general. Looks like spaces around are obviously used in various projects . In my opinion for more complex interpolations it looks better with spaces, but for basic variables interpolation I prefer no spaces. As you see there are various arguments for spaces (written in comments above), so maybe it's worth to discuss this rule.

**Comment by bbatsov** [comment]
This issue is not the place for such a discussion. Pretty much every rule gets questioned by someone, so nothing new here. :-) Let's implement the config option that was requested and be done with it.

**Comment by BM5k** [comment]
@bbatsov I think in general this _type_ of discussion is at the root of my love/hate relationship with rubocop. In particular, the coding styles used @devfu differ quite a bit from the “consensus”, in was that we’ve decided improve readability. The number of cops that don’t offer any configuration (other than on/off) mean that just about every rubocop update, we’re finding new cops to disable. Long term, this means our style isn’t actually being enforced, and discourages us from using rubocop in general.

I agree that this isn’t the place for this discussion, but where _is_ that place?

**Comment by bbatsov** [comment]
> The number of cops that don’t offer any configuration (other than on/off) mean that just about every rubocop update, we’re finding new cops to disable.

We have to start from somewhere and we usually start just from the most popular options. The project is always evolving and no popular convention was ever shot down. I'm extremely busy, however, and simply don't have the time to implement every option myself, so it's often up to the RuboCop users to extend it in the ways they want to.

> I agree that this isn’t the place for this discussion, but where is that place?

Either a dedicated issue for the general problem or our mailing list/gitter channel.

**Comment by gotrevor** [comment]
Regarding this comment: https://github.com/bbatsov/rubocop/issues/1971#issuecomment-112694828
> Even rubocop source contains spaces around string interpolation

I've corrected these three occurrences in my PR (#) to add this `space` configurable option.

However, there was not a bug in this cop.  These instances of spaces around interpolation were only in single-quoted strings used in one different spec.
